Pal 877 policy service documentation

developer01189998819991197253 opened the pull request.

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # Policy Service
 
-<span style="color:red">**Note:** As noted in the [documentation root](../README.md) Palisade is in an early stage of development, therefore the precise APIs contained in these service modules are likely to adapt as the project matures.</span>
+The responsibilities of the Policy Service is to provide the set of rules (filters or transformations) that need to be
```
I think it is responsibility.  This is describing one role and so it is singular.

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 --->
```
```suggestion

--->
```

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 --->
+# <img src="../logos/logo.svg" width="180">
```
```suggestion

# <img src="../logos/logo.svg" width="180">
```

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 --->
+# <img src="../logos/logo.svg" width="180">
+
+
```
```suggestion
```

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+| context                 | context                  | context           |
+| user                    | user                     | exception         |
+| resource                | resource                 | serverMetadata    |
+|                         | rule                     |                   |
```
```suggestion
|                         | rules                    |                   |
```

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+(fields marked with * are acquired from headers metadata)
+
+The service accepts a `PolicyRequest` and a token, then checks for rules, or policies in the cache; the technology of which is chosen in the Policy Services application.yaml
```
Would help to describe a token as it is not defined or self-explanatory.

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+The service accepts a `PolicyRequest` and a token, then checks for rules, or policies in the cache; the technology of which is chosen in the Policy Services application.yaml
+and Spring profiles. First, the Policy Service looks for resource level rules, these are rules that can be applied to any part of the resource path, and could limit what resources 
+the User has access to. The Policy Service recursively checks the parents of the resource for any record level rules, and applies them to the resource before continuing.
```
Understood that the convention with readme.mf files was to have each sentence on one line.  This looks like this practice is not being use here.

developer01189998819991197253 replied:
Theres no real convention. Each browser will format the Readme appropriatley and wrap lines.

m78233 commented on policy-service/README.md:
```diff
@@ -13,17 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+The service accepts a `PolicyRequest` and a token, then checks for rules, or policies in the cache; the technology of which is chosen in the Policy Services application.yaml
+and Spring profiles. First, the Policy Service looks for resource level rules, these are rules that can be applied to any part of the resource path, and could limit what resources 
+the User has access to. The Policy Service recursively checks the parents of the resource for any record level rules, and applies them to the resource before continuing.
+If no record or resource level rules are found for this resource, a `NoSuchPolicyException` is thrown. This is then packaged in a `AuditErrorMessage` object, and sent, with the `PolicyRequest` 
```
This gives the idea that NoSuchPolicyException equates to a AuditErrorMessage.  Suggest describing AuditErrorMessage in the wider context of any error in either the querying the resource or record level rules.

m78233 commented on policy-service/src/component-tests/java/uk/gov/gchq/palisade/component/policy/model/PolicyResponseTest.java:
```diff
@@ -15,53 +15,37 @@
-     * @throws IOException throws if the {@link PolicyResponse} object cannot be converted to a JsonContent.
-     *                     This equates to a failure to serialise or deserialize the string.
+     * @throws JsonProcessingException throws if the {@link PolicyResponse} object cannot be converted to a JsonContent.
+     *                                 This equates to a failure to serialise or deserialize the string.
```
```suggestion
     *                                 This equates to a failure to serialise or deserialise the string.
```

m78233 commented on policy-service/src/component-tests/java/uk/gov/gchq/palisade/component/policy/model/PolicyRequestTest.java:
```diff
@@ -15,53 +15,37 @@
-     * @throws IOException throws if the {@link PolicyRequest} object cannot be converted to a JsonContent.
-     *                     This equates to a failure to serialise or deserialize the string.
+     * @throws JsonProcessingException throws if the {@link PolicyRequest} object cannot be converted to a JsonContent.
+     *                                 This equates to a failure to serialise or deserialize the string.
```
```suggestion
     *                                 This equates to a failure to serialise or deserialise the string.
```

m78233 commented on policy-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/policy/kafka/KafkaTestConfiguration.java:
```diff
@@ -137,25 +136,25 @@ static String getKafkaBrokers(final KafkaContainer kafka) {
     }
 
     // Serializer for upstream test input
```
```suggestion
    // Serialiser for upstream test input
```

m78233 commented on policy-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/policy/kafka/KafkaTestConfiguration.java:
```diff
@@ -137,25 +136,25 @@ static String getKafkaBrokers(final KafkaContainer kafka) {
         }
     }
 
     // Deserializer for downstream test output
```
```suggestion
    // Deserialiser for downstream test output
```

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/StreamComponents.java:
```diff
@@ -49,27 +48,27 @@
      *
-     * @param system          the application's actor system to use to load config values
+     * @param system          the application's actor system used to load config values
      * @param keySerializer   the stream's key serialiser
```
```suggestion
     * @param keySerialiser   the stream's key serialiser
```

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/StreamComponents.java:
```diff
@@ -49,27 +48,27 @@
-     * @param system          the application's actor system to use to load config values
+     * @param system          the application's actor system used to load config values
      * @param keySerializer   the stream's key serialiser
      * @param valueSerializer the stream's value serialiser
```
```suggestion
     * @param valueSerialiser the stream's value serialiser
```

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/StreamComponents.java:
```diff
@@ -49,27 +48,27 @@
      * @param valueSerializer the stream's value serialiser
      * @return a {@link ProducerSettings} object for creating Akka {@link Sink}s
      */
     public ProducerSettings<K, V> producerSettings(final ActorSystem system, final Serializer<K> keySerializer, final Serializer<V> valueSerializer) {
```
```suggestion
    public ProducerSettings<K, V> producerSettings(final ActorSystem system, final Serializer<K> keySerialiser, final Serializer<V> valueSerialiser) {
```

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/StreamComponents.java:
```diff
@@ -49,27 +48,27 @@
      *
-     * @param system            the application's actor system to use to load config values
+     * @param system            the application's actor system used to load config values
      * @param keyDeserializer   the stream's key deserialiser
```
```suggestion
     * @param keyDeserialiser   the stream's key deserialiser
```

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/StreamComponents.java:
```diff
@@ -49,27 +48,27 @@
-     * @param system            the application's actor system to use to load config values
+     * @param system            the application's actor system used to load config values
      * @param keyDeserializer   the stream's key deserialiser
      * @param valueDeserializer the stream's value deserialiser
```
```suggestion
     * @param valueDeserialiser the stream's value deserialiser
```

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/StreamComponents.java:
```diff
@@ -49,27 +48,27 @@
      * @param valueDeserializer the stream's value deserialiser
      * @return a {@link ProducerSettings} object for creating Akka {@link Source}s
      */
     public ConsumerSettings<K, V> consumerSettings(final ActorSystem system, final Deserializer<K> keyDeserializer, final Deserializer<V> valueDeserializer) {
```
```suggestion
    public ConsumerSettings<K, V> consumerSettings(final ActorSystem system, final Deserializer<K> keyDeserialiser, final Deserializer<V> valueDeserialiser) {
```

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/config/StdPolicyPrepopulationFactory.java:
```diff
@@ -71,6 +71,7 @@ public StdPolicyPrepopulationFactory(final String resourceId, final Map<String,
         this.recordRules = recordRules;
     }
 
+    @SuppressWarnings({"java:S3740", "rawtypes"}) //Suppress parameterized type for generic class smell
```
Having trouble getting the comment to be on line 66, but the suggestion is to change "data-service" to Data Service.

m78233 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/model/PolicyResponse.java:
```diff
@@ -149,7 +150,7 @@ public static IUserId create() {
 
         /**
-         * Starter method for the Builder class that uses a PolicyRequest and appends the Rules.
+         * Starter method for the Builder class which uses a PolicyRequest and appends the Rules.
```
Can't get the comment to focus on the line I want, but on line 157 suggest changing policy-service to Policy Service

m78233 reviewed (changes requested):
Minor changes, main focus was on the readme and read over the rest with the lite touch.

m78233 reviewed (dismissed):
Looks good to me.

dev930018 commented on policy-service/README.md:
```diff
@@ -13,17 +13,166 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+(fields marked with * are acquired from headers metadata)
+
+The service accepts a `PolicyRequest` and a token created by the [Palisade Service](../palisade-serivce), then checks for rules, or policies in the cache; the technology of which is chosen in the Policy Services application.yaml
+and Spring profiles. 
+First, the Policy Service looks for resource level rules, these are rules that can be applied to any part of the resource path, and could limit what resources 
+the User has access to. The Policy Service recursively checks the parents of the resource for any record level rules, and applies them to the resource before continuing.
```
Prefer semantic line breaks in markdown rather than fixed-width - i.e. line-break every sentence, not every n characters

dev930018 commented on policy-service/README.md:
```diff
@@ -13,17 +13,166 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+(fields marked with * are acquired from headers metadata)
+
+The service accepts a `PolicyRequest` and a token created by the [Palisade Service](../palisade-serivce), then checks for rules, or policies in the cache; the technology of which is chosen in the Policy Services application.yaml
```
```suggestion
The service accepts a `PolicyRequest` and a token created by the [Palisade Service](../palisade-serivce), then checks for rules, or policies in the cache; the technology of which is chosen in the Policy Service's application.yaml
```

dev930018 commented on policy-service/README.md:
```diff
@@ -13,17 +13,166 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+The service accepts a `PolicyRequest` and a token created by the [Palisade Service](../palisade-serivce), then checks for rules, or policies in the cache; the technology of which is chosen in the Policy Services application.yaml
+and Spring profiles. 
+First, the Policy Service looks for resource level rules, these are rules that can be applied to any part of the resource path, and could limit what resources 
```
Some mention of the coarse-grain vocabulary distinction here

developer01189998819991197253 replied:
added

dev930018 commented on policy-service/README.md:
```diff
@@ -13,17 +13,166 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+If no record or resource level rules are found for this resource, a `NoSuchPolicyException` is thrown.
+An `AuditErrorMessage` is then built, containing the user information, context, resourceId, the error thrown, and any additional attributes. 
+This error object is then sent with the `PolicyRequest` to the [Audit service](../audit-service) to be audited appropriately.
+If however, resource level rules are found, and when applied against the resource, the resource has not been redacted, then the Policy Service will check the cache for Record Level rules. 
```
```suggestion
If, however, resource level rules are found, and when applied against the resource, the resource has not been redacted, then the Policy Service will check the cache for Record Level rules. 
```
leaving aside usage of 'however = but' vs 'however = no matter how' or 'however = in whatever manner'

dev930018 commented on policy-service/README.md:
```diff
@@ -13,17 +13,166 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+An `AuditErrorMessage` is then built, containing the user information, context, resourceId, the error thrown, and any additional attributes. 
+This error object is then sent with the `PolicyRequest` to the [Audit service](../audit-service) to be audited appropriately.
+If however, resource level rules are found, and when applied against the resource, the resource has not been redacted, then the Policy Service will check the cache for Record Level rules. 
+Record Level Rules are rules that can be applied against the data itself, and based off context can limit what the user can see in a particular resource.
```
Mention of fine-grained here maybe?
These two (fine and coarse) need only be mentioned as a sort of `..., sometimes referred to as 'fine-grained', ...`

developer01189998819991197253 replied:
added above

dev930018 commented on policy-service/README.md:
````diff
@@ -13,17 +13,166 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  }
+}'
+```
+## Example JSON Kafka Topic ('rule') Output
````
```suggestion

## Example JSON Kafka Topic ('rule') Output
```

dev930018 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/config/PolicyConfiguration.java:
```diff
@@ -29,7 +29,7 @@
      * @return a {@link List} of the objects that have implemented {@link PolicyPrepopulationFactory}.
      */
-    @SuppressWarnings("java:S1452")
+    @SuppressWarnings("java:S1452") // Suppress generic wildcard small
```
```suggestion
    @SuppressWarnings("java:S1452") // Allow generics wildcard
```

developer01189998819991197253 committed:
PAL-877 - Updated policy-service/README.md, added json example data, summary of service, objects in the service. Added .drawio diagrams of service in doc/README.md.
Updated javadoc in the rest of Policy-Service, updated test assertions, fixed grammar and spelling.
Removed redundant code, fixed code smells, replaced types with var where types are easily inferred.
Fixed json example data in user-service/README.md and attribute-masking-service/README.md
